lib/astutils.cpp

```diff
@@ -949,6 +949,8 @@ const Token *findLambdaEndToken(const Token *first)
         tok = tok->next();
     if (tok && tok->str() == "mutable")
         tok = tok->next();
+    if (tok && tok->originalName() == "->")
```

Comment:
This code does not use the ast. It seems we can clean it up.

For a lambda, the `astOperand1` for the `( ` token is set to `{` so I am hoping you can just use that. Something like:

    if (tok->str() == "[" && tok->astOperand1()->str() == "(" && tok->astOperand1()->astOperand1() == "{")
        return tok->astOperand1()->astOperand1()->link();

... please clean it up :-)

The testcases for ast handling of lambdas are in `TestTokenizer::astlambda`.. feel free to review that.

Reply:
Using `astOperand1` works like a charm, almost. It handles everything but lambdas that are specified as mutable, i.e. `[]() mutable {}`. Apart from that, it's much cleaner, and handles explicit return types regardless of the number of tokens involved (i.e. `const char *`, which my first approach didn't handle). I'm guessing `mutable` should be removed or ignored when dealing with lambdas?

Reply:
hmm.. then I wonder if you can fix handling of `mutable` also. the code is here:
https://github.com/danmar/cppcheck/blob/master/lib/tokenlist.cpp#L681
I would personally do that in a separate commit. If you change this code so it uses the AST (and you ignore mutable for a moment) I can merge it and your checker directly.
but if you rather fix mutable directly in this PR you can do it.